File: repoScript/Tosca/scenarioHackathon/clients/archive/upload_tosca_files_complete.py
#!/usr/bin/env python3
"""
OptimusDB TOSCA Upload Script (Python)
======================================
Uploads TOSCA YAML files to OptimusDB with base64 encoding
and persists template IDs to a JSON file.

Project: OptimusDB - EU Horizon Europe Grant 101135012

Usage:
    python upload_tosca_files_complete.py [base_url] [files_directory]

Example:
    python upload_tosca_files_complete.py http://localhost:18001 ./tosca_samples
"""

import sys
import os
import json
import base64
import requests
from datetime import datetime, timezone
from pathlib import Path
from typing import Dict, List, Optional, Tuple
from dataclasses import dataclass, asdict

# Color codes for terminal output
class Colors:
    RED = '\033[0;31m'
    GREEN = '\033[0;32m'
    YELLOW = '\033[1;33m'
    CYAN = '\033[0;36m'
    GRAY = '\033[0;90m'
    NC = '\033[0m'  # No Color
    BOLD = '\033[1m'

@dataclass
class UploadResult:
    """Result of a TOSCA file upload"""
    filename: str
    description: str
    template_id: str
    queryable: bool
    storage_location: str
    uploaded_at: str

@dataclass
class UploadSession:
    """Summary of upload session"""
    timestamp: str
    base_url: str
    total_files: int
    uploaded: int
    failed: int

class ToscaUploader:
    """Main class for uploading TOSCA files to OptimusDB"""

    # TOSCA files configuration
    TOSCA_FILES = {
        "webapp_adt.yaml": "WebApp Microservices Application",
        "capacity_profile.yaml": "Edge Cluster Capacity Profile",
        "opentofu_hybrid.yaml": "Hybrid Infrastructure with OpenTofu",
        "deployment_plan.yaml": "Deployment Plan with Workflows",
        "app_requirements.yaml": "ML Training Application Requirements"
    }

    def __init__(self, base_url: str = "http://localhost:18001",
                 files_dir: str = "."):
        self.base_url = base_url
        self.files_dir = Path(files_dir)
        self.output_file = "uploaded_tosca_templates.json"
        self.log_file = f"upload_log_{datetime.now().strftime('%Y%m%d_%H%M%S')}.txt"

        self.uploaded_count = 0
        self.failed_count = 0
        self.upload_results: List[UploadResult] = []

    def log(self, message: str) -> None:
        """Write message to log file"""
        timestamp = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
        with open(self.log_file, 'a', encoding='utf-8') as f:
            f.write(f"[{timestamp}] {message}\n")

    def print_header(self, text: str) -> None:
        """Print section header"""
        print()
        print(f"{Colors.CYAN}{'═' * 63}{Colors.NC}")
        print(f"{Colors.CYAN}{text}{Colors.NC}")
        print(f"{Colors.CYAN}{'═' * 63}{Colors.NC}")
        print()

    def print_success(self, message: str) -> None:
        """Print success message"""
        print(f"{Colors.GREEN}✅ {message}{Colors.NC}")
        self.log(f"SUCCESS: {message}")

    def print_error(self, message: str) -> None:
        """Print error message"""
        print(f"{Colors.RED}❌ {message}{Colors.NC}")
        self.log(f"ERROR: {message}")

    def print_warning(self, message: str) -> None:
        """Print warning message"""
        print(f"{Colors.YELLOW}⚠️  {message}{Colors.NC}")
        self.log(f"WARNING: {message}")

    def print_info(self, message: str) -> None:
        """Print info message"""
        print(f"{Colors.CYAN}ℹ️  {message}{Colors.NC}")

    def print_detail(self, message: str) -> None:
        """Print detail message"""
        print(f"{Colors.GRAY}   {message}{Colors.NC}")

    def check_dependencies(self) -> bool:
        """Check if required dependencies are installed"""
        self.print_info("Checking dependencies...")

        try:
            import requests
            self.print_success("All dependencies available")
            return True
        except ImportError:
            self.print_error("Missing required dependency: requests")
            print("\nPlease install requests:")
            print("  pip install requests")
            return False

    def test_connectivity(self) -> bool:
        """Test connection to OptimusDB API"""
        self.print_info(f"Testing connection to {self.base_url}...")

        try:
            response = requests.get(
                f"{self.base_url}/health",
                timeout=5
            )
            self.print_success("API is reachable")
            return True
        except requests.exceptions.RequestException:
            self.print_warning("Health endpoint not responding (this may be normal)")
            self.print_info("Attempting to continue anyway...")
            return True

    def convert_to_base64(self, filepath: Path) -> Optional[str]:
        """Convert file to base64 encoding"""
        try:
            with open(filepath, 'rb') as f:
                content = f.read()
                return base64.b64encode(content).decode('utf-8')
        except Exception as e:
            self.print_error(f"Failed to convert file to base64: {e}")
            return None

    def upload_tosca_file(self, filename: str, description: str) -> Optional[UploadResult]:
        """Upload a single TOSCA file"""
        print()
        self.print_info(f"Processing: {description}")
        self.print_detail(f"File: {filename}")

        filepath = self.files_dir / filename

        # Check file exists
        if not filepath.exists():
            self.print_error(f"File not found: {filepath}")
            return None

        # Get file size
        size_kb = filepath.stat().st_size / 1024
        self.print_detail(f"Size: {size_kb:.2f} KB")

        # Convert to base64
        self.print_detail("Converting to base64...")
        base64_content = self.convert_to_base64(filepath)

        if not base64_content:
            return None

        # Prepare request payload
        payload = {
            "file": base64_content,
            "filename": filename,
            "store_full_structure": True
        }

        # Upload to OptimusDB
        self.print_detail(f"Uploading to {self.base_url}/swarmkb/upload...")

        try:
            response = requests.post(
                f"{self.base_url}/swarmkb/upload",
                json=payload,
                headers={"Content-Type": "application/json"},
                timeout=60
            )

            # Check HTTP status
            if response.status_code != 200:
                self.print_error(f"Upload failed with HTTP {response.status_code}")
                self.print_detail(f"Response: {response.text}")
                return None

            # Parse response
            data = response.json()

            if data.get('status') != 200:
                error_msg = data.get('message', 'Unknown error')
                self.print_error(f"Upload failed: {error_msg}")
                return None

            # Extract response data
            response_data = data.get('data', {})
            template_id = response_data.get('template_id')
            queryable = response_data.get('queryable', False)
            storage_location = response_data.get('storage_location', '')

            if not template_id:
                self.print_error("No template ID returned in response")
                self.print_detail(f"Response: {data}")
                return None

            # Success!
            self.print_success("Upload successful")
            self.print_detail(f"Template ID: {template_id}")
            self.print_detail(f"Queryable: {queryable}")
            self.print_detail(f"Storage: {storage_location}")

            # Create result object
            return UploadResult(
                filename=filename,
                description=description,
                template_id=template_id,
                queryable=queryable,
                storage_location=storage_location,
                uploaded_at=datetime.now(timezone.utc).isoformat()
            )

        except requests.exceptions.RequestException as e:
            self.print_error(f"Upload failed: {e}")
            return None
        except json.JSONDecodeError as e:
            self.print_error(f"Failed to parse response JSON: {e}")
            return None

    def save_results(self) -> None:
        """Save upload results to JSON file"""
        output_data = {
            "upload_session": asdict(UploadSession(
                timestamp=datetime.now(timezone.utc).isoformat(),
                base_url=self.base_url,
                total_files=len(self.TOSCA_FILES),
                uploaded=self.uploaded_count,
                failed=self.failed_count
            )),
            "templates": [asdict(result) for result in self.upload_results]
        }

        with open(self.output_file, 'w', encoding='utf-8') as f:
            json.dump(output_data, f, indent=2, ensure_ascii=False)

        self.print_success(f"Results saved to: {self.output_file}")

    def verify_uploads(self) -> bool:
        """Verify uploads by querying the database"""
        self.print_info("Verifying uploads...")

        query_payload = {
            "method": {"cmd": "crudget", "argcnt": 1},
            "dstype": "dsswres",
            "criteria": []
        }

        try:
            response = requests.post(
                f"{self.base_url}/swarmkb/command",
                json=query_payload,
                headers={"Content-Type": "application/json"},
                timeout=30
            )

            data = response.json()
            total_count = len(data.get('data', []))

            if total_count > 0:
                self.print_success(f"Verified: {total_count} total templates in database")

                # Count TOSCA templates
                tosca_count = sum(
                    1 for item in data.get('data', [])
                    if 'tosca_definitions_version' in item
                )
                self.print_detail(f"TOSCA templates: {tosca_count}")

                return True
            else:
                self.print_warning("Could not verify uploads (query returned no results)")
                return False

        except Exception as e:
            self.print_warning(f"Could not verify uploads: {e}")
            return False

    def run(self) -> int:
        """Main execution method"""
        self.print_header("OptimusDB TOSCA Upload Script")

        print("Configuration:")
        print(f"  Base URL: {self.base_url}")
        print(f"  Files Directory: {self.files_dir}")
        print(f"  Output File: {self.output_file}")
        print(f"  Log File: {self.log_file}")
        print()

        # Check dependencies
        if not self.check_dependencies():
            return 1

        # Test connectivity
        self.test_connectivity()

        # Process each file
        self.print_header("Uploading TOSCA Files")

        for filename, description in self.TOSCA_FILES.items():
            result = self.upload_tosca_file(filename, description)

            if result:
                self.upload_results.append(result)
                self.uploaded_count += 1
            else:
                self.failed_count += 1

            # Brief pause between uploads
            import time
            time.sleep(1)

        # Summary
        print()
        self.print_header("Upload Summary")

        total_files = len(self.TOSCA_FILES)
        print(f"Total Files:     {total_files}")
        print(f"{Colors.GREEN}Uploaded:        {self.uploaded_count}{Colors.NC}")
        print(f"{Colors.RED}Failed:          {self.failed_count}{Colors.NC}")
        print()

        # Save results if any succeeded
        if self.uploaded_count > 0:
            self.save_results()

            print()
            self.print_info(f"Template IDs saved to {self.output_file}")
            print()

            # Show uploaded templates
            print("Uploaded Templates:")
            for result in self.upload_results:
                print(f"{Colors.GREEN}  ✓ {result.description}{Colors.NC}")
                print(f"{Colors.GRAY}    ID: {result.template_id}{Colors.NC}")

            # Verify uploads
            print()
            self.verify_uploads()

        # Final status
        print()
        if self.failed_count == 0:
            self.print_header("✅ All Uploads Successful!")
            return 0
        elif self.uploaded_count > 0:
            self.print_header("⚠️  Partial Success - Some Uploads Failed")
            return 1
        else:
            self.print_header("❌ All Uploads Failed")
            return 1

def main():
    """Main entry point"""
    # Parse command line arguments
    base_url = sys.argv[1] if len(sys.argv) > 1 else "http://localhost:18001"
    files_dir = sys.argv[2] if len(sys.argv) > 2 else "."

    # Create uploader and run
    uploader = ToscaUploader(base_url, files_dir)
    exit_code = uploader.run()

    sys.exit(exit_code)

if __name__ == "__main__":
    main()

File: repoScript/Tosca/scenarioHackathon/clients/archive/upload_tosca_files_complete.go
package main

/*
OptimusDB TOSCA Upload Script (Go)
===================================
Uploads TOSCA YAML files to OptimusDB with base64 encoding
and persists template IDs to a JSON file.

Project: OptimusDB - EU Horizon Europe Grant 101135012

Usage:
	go run upload_tosca_files_complete.go [base_url] [files_directory]

Example:
	go run upload_tosca_files_complete.go http://localhost:18001 ./tosca_samples

Build:
	go build -o upload_tosca upload_tosca_files_complete.go

Dependencies:
	Standard library only - no external dependencies required!
*/

import (
	"bytes"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"time"
)

// Color codes for terminal output
const (
	ColorReset  = "\033[0m"
	ColorRed    = "\033[31m"
	ColorGreen  = "\033[32m"
	ColorYellow = "\033[33m"
	ColorCyan   = "\033[36m"
	ColorGray   = "\033[90m"
)

// ToscaFile represents a TOSCA file to upload
type ToscaFile struct {
	Filename    string
	Description string
}

// UploadRequest represents the request payload for uploading a TOSCA file
type UploadRequest struct {
	File               string `json:"file"`
	Filename           string `json:"filename"`
	StoreFullStructure bool   `json:"store_full_structure"`
}

// UploadResponseData represents the data field in upload response
type UploadResponseData struct {
	TemplateID      string `json:"template_id"`
	Queryable       bool   `json:"queryable"`
	StorageLocation string `json:"storage_location"`
	Filename        string `json:"filename"`
}

// UploadResponse represents the response from upload endpoint
type UploadResponse struct {
	Status  int                `json:"status"`
	Message string             `json:"message"`
	Data    UploadResponseData `json:"data"`
}

// UploadResult represents the result of uploading a file
type UploadResult struct {
	Filename        string `json:"filename"`
	Description     string `json:"description"`
	TemplateID      string `json:"template_id"`
	Queryable       bool   `json:"queryable"`
	StorageLocation string `json:"storage_location"`
	UploadedAt      string `json:"uploaded_at"`
}

// UploadSession represents the summary of the upload session
type UploadSession struct {
	Timestamp  string `json:"timestamp"`
	BaseURL    string `json:"base_url"`
	TotalFiles int    `json:"total_files"`
	Uploaded   int    `json:"uploaded"`
	Failed     int    `json:"failed"`
}

// OutputData represents the final JSON output
type OutputData struct {
	Session   UploadSession  `json:"upload_session"`
	Templates []UploadResult `json:"templates"`
}

// QueryRequest represents a query request
type QueryRequest struct {
	Method   map[string]interface{} `json:"method"`
	DsType   string                 `json:"dstype"`
	Criteria []interface{}          `json:"criteria"`
}

// QueryResponse represents a query response
type QueryResponse struct {
	Status int                      `json:"status"`
	Data   []map[string]interface{} `json:"data"`
}

// ToscaUploader manages the upload process
type ToscaUploader struct {
	BaseURL       string
	FilesDir      string
	OutputFile    string
	LogFile       string
	UploadedCount int
	FailedCount   int
	Results       []UploadResult
	LogWriter     *os.File
}

// TOSCA files configuration
var toscaFiles = []ToscaFile{
	{"webapp_adt.yaml", "WebApp Microservices Application"},
	{"capacity_profile.yaml", "Edge Cluster Capacity Profile"},
	{"opentofu_hybrid.yaml", "Hybrid Infrastructure with OpenTofu"},
	{"deployment_plan.yaml", "Deployment Plan with Workflows"},
	{"app_requirements.yaml", "ML Training Application Requirements"},
}

// NewToscaUploader creates a new uploader instance
func NewToscaUploader(baseURL, filesDir string) *ToscaUploader {
	timestamp := time.Now().Format("20060102_150405")
	logFile := fmt.Sprintf("upload_log_%s.txt", timestamp)

	logWriter, err := os.OpenFile(logFile, os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0644)
	if err != nil {
		fmt.Printf("Warning: Could not open log file: %v\n", err)
	}

	return &ToscaUploader{
		BaseURL:    baseURL,
		FilesDir:   filesDir,
		OutputFile: "uploaded_tosca_templates.json",
		LogFile:    logFile,
		Results:    make([]UploadResult, 0),
		LogWriter:  logWriter,
	}
}

// Close closes the log file
func (u *ToscaUploader) Close() {
	if u.LogWriter != nil {
		u.LogWriter.Close()
	}
}

// Log writes a message to the log file
func (u *ToscaUploader) Log(message string) {
	timestamp := time.Now().Format("2006-01-02 15:04:05")
	logMessage := fmt.Sprintf("[%s] %s\n", timestamp, message)
	if u.LogWriter != nil {
		u.LogWriter.WriteString(logMessage)
	}
}

// Print functions with colors
func (u *ToscaUploader) PrintHeader(text string) {
	fmt.Println()
	fmt.Printf("%s%s%s\n", ColorCyan, "═══════════════════════════════════════════════════════════", ColorReset)
	fmt.Printf("%s%s%s\n", ColorCyan, text, ColorReset)
	fmt.Printf("%s%s%s\n", ColorCyan, "═══════════════════════════════════════════════════════════", ColorReset)
	fmt.Println()
}

func (u *ToscaUploader) PrintSuccess(message string) {
	fmt.Printf("%s✅ %s%s\n", ColorGreen, message, ColorReset)
	u.Log("SUCCESS: " + message)
}

func (u *ToscaUploader) PrintError(message string) {
	fmt.Printf("%s❌ %s%s\n", ColorRed, message, ColorReset)
	u.Log("ERROR: " + message)
}

func (u *ToscaUploader) PrintWarning(message string) {
	fmt.Printf("%s⚠️  %s%s\n", ColorYellow, message, ColorReset)
	u.Log("WARNING: " + message)
}

func (u *ToscaUploader) PrintInfo(message string) {
	fmt.Printf("%sℹ️  %s%s\n", ColorCyan, message, ColorReset)
}

func (u *ToscaUploader) PrintDetail(message string) {
	fmt.Printf("%s   %s%s\n", ColorGray, message, ColorReset)
}

// TestConnectivity tests connection to the API
func (u *ToscaUploader) TestConnectivity() bool {
	u.PrintInfo(fmt.Sprintf("Testing connection to %s...", u.BaseURL))

	client := &http.Client{Timeout: 5 * time.Second}
	resp, err := client.Get(u.BaseURL + "/health")

	if err == nil && resp.StatusCode == 200 {
		u.PrintSuccess("API is reachable")
		resp.Body.Close()
		return true
	}

	u.PrintWarning("Health endpoint not responding (this may be normal)")
	u.PrintInfo("Attempting to continue anyway...")
	return true
}

// ConvertToBase64 converts a file to base64
func (u *ToscaUploader) ConvertToBase64(filepath string) (string, error) {
	data, err := os.ReadFile(filepath)
	if err != nil {
		return "", err
	}

	encoded := base64.StdEncoding.EncodeToString(data)
	return encoded, nil
}

// UploadToscaFile uploads a single TOSCA file
func (u *ToscaUploader) UploadToscaFile(file ToscaFile) *UploadResult {
	fmt.Println()
	u.PrintInfo(fmt.Sprintf("Processing: %s", file.Description))
	u.PrintDetail(fmt.Sprintf("File: %s", file.Filename))

	filepath := filepath.Join(u.FilesDir, file.Filename)

	// Check file exists
	fileInfo, err := os.Stat(filepath)
	if err != nil {
		u.PrintError(fmt.Sprintf("File not found: %s", filepath))
		return nil
	}

	// Get file size
	sizeKB := float64(fileInfo.Size()) / 1024.0
	u.PrintDetail(fmt.Sprintf("Size: %.2f KB", sizeKB))

	// Convert to base64
	u.PrintDetail("Converting to base64...")
	base64Content, err := u.ConvertToBase64(filepath)
	if err != nil {
		u.PrintError(fmt.Sprintf("Failed to convert to base64: %v", err))
		return nil
	}

	// Prepare request
	uploadReq := UploadRequest{
		File:               base64Content,
		Filename:           file.Filename,
		StoreFullStructure: true,
	}

	jsonData, err := json.Marshal(uploadReq)
	if err != nil {
		u.PrintError(fmt.Sprintf("Failed to marshal JSON: %v", err))
		return nil
	}

	// Upload to OptimusDB
	u.PrintDetail(fmt.Sprintf("Uploading to %s/swarmkb/upload...", u.BaseURL))

	client := &http.Client{Timeout: 60 * time.Second}
	resp, err := client.Post(
		u.BaseURL+"/swarmkb/upload",
		"application/json",
		bytes.NewBuffer(jsonData),
	)

	if err != nil {
		u.PrintError(fmt.Sprintf("Upload failed: %v", err))
		return nil
	}
	defer resp.Body.Close()

	// Check HTTP status
	if resp.StatusCode != 200 {
		body, _ := io.ReadAll(resp.Body)
		u.PrintError(fmt.Sprintf("Upload failed with HTTP %d", resp.StatusCode))
		u.PrintDetail(fmt.Sprintf("Response: %s", string(body)))
		return nil
	}

	// Parse response
	var uploadResp UploadResponse
	if err := json.NewDecoder(resp.Body).Decode(&uploadResp); err != nil {
		u.PrintError(fmt.Sprintf("Failed to parse response: %v", err))
		return nil
	}

	if uploadResp.Status != 200 {
		u.PrintError(fmt.Sprintf("Upload failed: %s", uploadResp.Message))
		return nil
	}

	if uploadResp.Data.TemplateID == "" {
		u.PrintError("No template ID returned in response")
		return nil
	}

	// Success!
	u.PrintSuccess("Upload successful")
	u.PrintDetail(fmt.Sprintf("Template ID: %s", uploadResp.Data.TemplateID))
	u.PrintDetail(fmt.Sprintf("Queryable: %t", uploadResp.Data.Queryable))
	u.PrintDetail(fmt.Sprintf("Storage: %s", uploadResp.Data.StorageLocation))

	// Create result
	return &UploadResult{
		Filename:        file.Filename,
		Description:     file.Description,
		TemplateID:      uploadResp.Data.TemplateID,
		Queryable:       uploadResp.Data.Queryable,
		StorageLocation: uploadResp.Data.StorageLocation,
		UploadedAt:      time.Now().UTC().Format(time.RFC3339),
	}
}

// SaveResults saves results to JSON file
func (u *ToscaUploader) SaveResults() error {
	outputData := OutputData{
		Session: UploadSession{
			Timestamp:  time.Now().UTC().Format(time.RFC3339),
			BaseURL:    u.BaseURL,
			TotalFiles: len(toscaFiles),
			Uploaded:   u.UploadedCount,
			Failed:     u.FailedCount,
		},
		Templates: u.Results,
	}

	jsonData, err := json.MarshalIndent(outputData, "", "  ")
	if err != nil {
		return err
	}

	err = os.WriteFile(u.OutputFile, jsonData, 0644)
	if err != nil {
		return err
	}

	u.PrintSuccess(fmt.Sprintf("Results saved to: %s", u.OutputFile))
	return nil
}

// VerifyUploads queries the database to verify uploads
func (u *ToscaUploader) VerifyUploads() bool {
	u.PrintInfo("Verifying uploads...")

	queryReq := QueryRequest{
		Method: map[string]interface{}{
			"cmd":    "crudget",
			"argcnt": 1,
		},
		DsType:   "dsswres",
		Criteria: make([]interface{}, 0),
	}

	jsonData, err := json.Marshal(queryReq)
	if err != nil {
		u.PrintWarning(fmt.Sprintf("Could not verify uploads: %v", err))
		return false
	}

	client := &http.Client{Timeout: 30 * time.Second}
	resp, err := client.Post(
		u.BaseURL+"/swarmkb/command",
		"application/json",
		bytes.NewBuffer(jsonData),
	)

	if err != nil {
		u.PrintWarning(fmt.Sprintf("Could not verify uploads: %v", err))
		return false
	}
	defer resp.Body.Close()

	var queryResp QueryResponse
	if err := json.NewDecoder(resp.Body).Decode(&queryResp); err != nil {
		u.PrintWarning(fmt.Sprintf("Could not parse query response: %v", err))
		return false
	}

	totalCount := len(queryResp.Data)

	if totalCount > 0 {
		u.PrintSuccess(fmt.Sprintf("Verified: %d total templates in database", totalCount))

		// Count TOSCA templates
		toscaCount := 0
		for _, item := range queryResp.Data {
			if _, ok := item["tosca_definitions_version"]; ok {
				toscaCount++
			}
		}

		u.PrintDetail(fmt.Sprintf("TOSCA templates: %d", toscaCount))
		return true
	}

	u.PrintWarning("Could not verify uploads (query returned no results)")
	return false
}

// Run executes the upload process
func (u *ToscaUploader) Run() int {
	u.PrintHeader("OptimusDB TOSCA Upload Script")

	fmt.Println("Configuration:")
	fmt.Printf("  Base URL: %s\n", u.BaseURL)
	fmt.Printf("  Files Directory: %s\n", u.FilesDir)
	fmt.Printf("  Output File: %s\n", u.OutputFile)
	fmt.Printf("  Log File: %s\n", u.LogFile)
	fmt.Println()

	// Test connectivity
	u.TestConnectivity()

	// Process each file
	u.PrintHeader("Uploading TOSCA Files")

	for _, file := range toscaFiles {
		result := u.UploadToscaFile(file)

		if result != nil {
			u.Results = append(u.Results, *result)
			u.UploadedCount++
		} else {
			u.FailedCount++
		}

		// Brief pause between uploads
		time.Sleep(1 * time.Second)
	}

	// Summary
	fmt.Println()
	u.PrintHeader("Upload Summary")

	totalFiles := len(toscaFiles)
	fmt.Printf("Total Files:     %d\n", totalFiles)
	fmt.Printf("%sUploaded:        %d%s\n", ColorGreen, u.UploadedCount, ColorReset)
	fmt.Printf("%sFailed:          %d%s\n", ColorRed, u.FailedCount, ColorReset)
	fmt.Println()

	// Save results if any succeeded
	if u.UploadedCount > 0 {
		if err := u.SaveResults(); err != nil {
			u.PrintError(fmt.Sprintf("Failed to save results: %v", err))
		}

		fmt.Println()
		u.PrintInfo(fmt.Sprintf("Template IDs saved to %s", u.OutputFile))
		fmt.Println()

		// Show uploaded templates
		fmt.Println("Uploaded Templates:")
		for _, result := range u.Results {
			fmt.Printf("%s  ✓ %s%s\n", ColorGreen, result.Description, ColorReset)
			fmt.Printf("%s    ID: %s%s\n", ColorGray, result.TemplateID, ColorReset)
		}

		// Verify uploads
		fmt.Println()
		u.VerifyUploads()
	}

	// Final status
	fmt.Println()
	if u.FailedCount == 0 {
		u.PrintHeader("✅ All Uploads Successful!")
		return 0
	} else if u.UploadedCount > 0 {
		u.PrintHeader("⚠️  Partial Success - Some Uploads Failed")
		return 1
	} else {
		u.PrintHeader("❌ All Uploads Failed")
		return 1
	}
}

func main() {
	// Parse command line arguments
	baseURL := "http://localhost:18001"
	filesDir := "."

	if len(os.Args) > 1 {
		baseURL = os.Args[1]
	}
	if len(os.Args) > 2 {
		filesDir = os.Args[2]
	}

	// Create uploader
	uploader := NewToscaUploader(baseURL, filesDir)
	defer uploader.Close()

	// Run and exit with appropriate code
	exitCode := uploader.Run()
	os.Exit(exitCode)
}
